spring-kafka/src/main/java/org/springframework/kafka/listener/AbstractMessageListenerContainer.java

```diff
@@ -475,6 +475,28 @@ public void setRecordInterceptor(@Nullable RecordInterceptor<K, V> recordInterce
 		this.recordInterceptor = recordInterceptor;
 	}
 
+	/**
+	 * Add an interceptor to be called before calling the record listener
+	 * If the {@link AbstractMessageListenerContainer} is not configured with a
+	 * {@link CompositeRecordInterceptor}, {@link CompositeRecordInterceptor} will be created internally
+	 * and configured to hold the added interceptor.
+	 * Does not apply to batch listeners.
+	 * @param recordInterceptor the interceptor.
+	 * @since 4.0
+	 */
+	public void addRecordInterceptor(RecordInterceptor<K, V> recordInterceptor) {
+		if (this.recordInterceptor instanceof CompositeRecordInterceptor<K, V> compositeRecordInterceptor) {
+			compositeRecordInterceptor.addRecordInterceptor(recordInterceptor);
```

Comment:
So, if the composite is an external and shared, calling this from one container would mutate all others. I understand a desire to have everything container related in one place, but consequences might not be good.
Or we just leave responsibility only in the composite, or bite a bullet and always have a composite here internally as a value of interceptor property. This way any add on the container would not effect the rest of application . The awkward situation when composite is provided, but I don’t see any other way to protect from race condition .
Of course, I’d prefer to have add only in the composite, but I’m opened for any other arguments.